### PR TITLE
Changes to episodes 3 and 4

### PR DESCRIPTION
**Pull request by ALuesink**

### CODE REVIEW

File: _episodes/03-dependency_graphs.md [hunk 1]
```diff
@@ -8,21 +8,162 @@ questions:
...
+> ## Learning objectives
+>By the end of this episode, learners should be able to __understand the flow of data between tools__ 
+> and __explain that a workflow is a dependency graph__
+>and __sketch their workflow, both by hand, and with an automated visualizer__
+>and __recognise that workflow development can be iterative i.e. that it doesn't have to happen all at once__.
+{: .callout}
+
+
+## Multi-Step Workflow
+In the previous episode a single step workflow was shown. To make a multi-step workflow, you extend the `steps` field.
```
Comment (mr-c) [inline _episodes/03-dependency_graphs.md]:
```suggestion
In the previous episode a single step workflow was shown. To make a multi-step workflow, you add more entries to the `steps` field.
```

File: _episodes/03-dependency_graphs.md [hunk 2]
```diff
@@ -8,21 +8,162 @@ questions:
...
+In this episode, the workflow is extended with the next 2 steps of the RNA-sequencing analysis.
+The next 2 steps are alingment of the reads and indexing the alignment. In this example `STAR` and `samtools` are used for these tasks.
```
Comment (mr-c) [inline _episodes/03-dependency_graphs.md]:
```suggestion
In this episode, the workflow is extended with the next two steps of the RNA-sequencing analysis.
The next two steps are alignment of the reads and indexing the alignments. We will be using the `STAR` and `samtools` tools for these tasks.
```
And please add links to the STAR & samtools entries on https://bio.tools

File: _episodes/03-dependency_graphs.md [hunk 3]
```diff
@@ -8,21 +8,162 @@ questions:
...
+      bam_sorted: mapping_reads/alignment
+    out: [bam_sorted_indexed]
+
+outputs: 
+  qc_html:
+    type: File
+	outputSource: quality_control/html_file
+  bam_sorted_indexed:
+    type: File
+	outputSource: index_alignment/bam_sorted_indexed
```
Comment (mr-c) [inline _episodes/03-dependency_graphs.md]:
```suggestion
outputs: 
  qc_html:
    type: File
    outputSource: quality_control/html_file
  bam_sorted_indexed:
    type: File
    outputSource: index_alignment/bam_sorted_indexed
```

File: _episodes/03-dependency_graphs.md [hunk 4]
```diff
@@ -8,21 +8,162 @@ questions:
...
+  quality_control:
+    run: bio-cwl-tools/fastqc/fastqc_2.cwl
+	in:
+	  reads_file: rna_reads_human
+    out: [html_file]
+	
+  mapping_reads:
+    requirements:
+	  ResourceRequirement:
+	    ramMin: 9000
```
Comment (mr-c) [inline _episodes/03-dependency_graphs.md]:
```suggestion
    requirements:
	    ResourceRequirement:
	      ramMin: 9000
```

File: _episodes/03-dependency_graphs.md [hunk 5]
```diff
@@ -8,21 +8,162 @@ questions:
...
+	outputSource: index_alignment/bam_sorted_indexed
+~~~
+{: .language-yaml}
+
+The workflow file shows the first 3 steps of the RNA-seq analysis: `quality_control`, `mapping_reads` and `index_alignment`.
+The `index_alignment` step uses the `alignment` output of the `mapping_reads` step. 
+You do this by referencing the output of the `mapping_reads` step in the `in` field of the `index_alignment` step.
+This is similar to referencing the outputs of the different steps in the `outputs` section. 
+
+The `mapping_reads` step needs some extra information which are command line arguments if you run the STAR aligner on the command line.
```
Comment (mr-c) [inline _episodes/03-dependency_graphs.md]:
```suggestion
The `mapping_reads` step needs some extra information beyond the inputs from the other steps, which is done by providing `default` values. If you want, you can read the `bio-cwl-tools/STAR/STAR-Align.cwl` file to see how these extra inputs are transformed into command line options to the `STAR` program.
```

File: _episodes/03-dependency_graphs.md [hunk 6]
```diff
@@ -8,21 +8,162 @@ questions:
...
+This information is provided in the `in` field. To run the tool better, it needs more RAM than the default.
+In the `requirements` field the `ResourceRequirement` field allocates a minimum of 9000 MB of RAM.
```
Comment (mr-c) [inline _episodes/03-dependency_graphs.md]:
```suggestion
This information is provided in the `in` field.

To run the tool better, it needs more RAM than the default. So there is a `requirements` entry inside the `mapping_reads` step definition with a `ResourceRequirement` to allocate a minimum of 9000 MiB of RAM.
```

File: _episodes/03-dependency_graphs.md [hunk 7]
```diff
@@ -8,21 +8,162 @@ questions:
...
+
+The newly added steps also need more inputs. The `mapping_reads` step needs a directory that contains the reference genome necessary for the mapping.
+This is added in the `inputs` field and in the YAML input file, `workflow_input.yml`.
```
Comment (mr-c) [inline _episodes/03-dependency_graphs.md]:
```suggestion
The newly added `mapping_reads` step also need an input not provided by any of our other steps, therefore an additional workflow-level input is added: a directory that contains the reference genome necessary for the mapping.

This `ref_genome` is added in the `inputs` field of the workflow and in the YAML input file, `workflow_input.yml`.
```

File: _episodes/03-dependency_graphs.md [hunk 8]
```diff
@@ -8,21 +8,162 @@ questions:
...
+> ## Iterative working
+> Working on a workflow is often not something that happens all at once. 
+> Sometimes you already have a shell script ready that can be converted to a CWL workflow. 
+> Other times it is similar to this tutorial, you start with a single-step workflow and extend it to a multi-step workflow.
+> This is all iterative working, a continuous work in progress.
+{: . callout}
+
+## Visualising a workflow
+
+A workflow is actually a directed acyclic graph (DAG). This means that the workflow has a certain direction, from inputs to outputs, and has no cycles.
```
Comment (mr-c) [inline _episodes/03-dependency_graphs.md]:
```suggestion
A CWL workflow is a directed acyclic graph (DAG). This means that
1) The workflow has a certain direction, from workflow inputs to step inputs, from step outputs to other step inputs, and from step outputs to workflow outputs
and
2) The workflow definition has no cycles.
```

File: _episodes/03-dependency_graphs.md [hunk 9]
```diff
@@ -8,21 +8,162 @@ questions:
...
+A CWL workflow is a dependency graph. Each step in the workflow depends on either an input or another step.
```
Comment (mr-c) [inline _episodes/03-dependency_graphs.md]:
```suggestion
A CWL workflow is a dependency graph. Each input for a step in the workflow depends on either a workflow-level input or a particular output from another step.
```

File: _episodes/03-dependency_graphs.md [hunk 10]
```diff
@@ -8,21 +8,162 @@ questions:
...
+
+It is also possible to generate the graph in the command line. `cwltool` has a function that makes a graph. 
+The `--print-dot` option will print a file suitable for Graphviz `dot` program. This is the command to generate a Scalable Vector Graphic (SVG) file:
+
+~~~
+cwltool --print-dot rna_seq_workflow.cwl | dot -Tsvg > workflow_graph.svg
+~~~
+{: .language-bash}
+
+The resulting SVG file displays the same graph as the one in the cwlviewer. The SVG file can be opened in a internetbrowser and in [Inkscape](https://inkscape.org/)
```
Comment (mr-c) [inline _episodes/03-dependency_graphs.md]:
```suggestion
The resulting SVG file displays the same graph as the one in the cwlviewer. The SVG file can be opened in any web browser and in [Inkscape](https://inkscape.org/), for example.
```

**Review by mr-c — CHANGES_REQUESTED** [review]
Just a review on episode 3; waiting for additional episode 4 changes to come in before reviewing (as requested)

File: _episodes/03-dependency_graphs.md (hunk 2 above)
Comment (mr-c) [reply _episodes/03-dependency_graphs.md]: https://bio.tools/star
https://bio.tools/samtools

File: _episodes/04-reusing_tools.md [hunk 11]
```diff
@@ -18,35 +18,20 @@ keypoints:
...
 
+The CWL documents of the previous steps were already provided for you, however, you can also find them in this library.
+In this episode you will use the bio-cwl-tools library to add the last step to the workflow. 
 
 ## Adding new step in workflow
-The last step of the workflow is counting of the reads. For this step the `featureCounts` step will be used.
-To implement the last step the bio-cwl-tools library library will be used.
-
+The first step is finding the tool you want to use in the [bio-cwl-tools](https://github.com/common-workflow-library/bio-cwl-tools).
+The last step of our workflow is counting the RNA-seq reads for which we will use the [`featureCounts`](https://bio.tools/featurecounts) tool.
```
Comment (mr-c) [inline _episodes/04-reusing_tools.md]:
```suggestion
The last step of our workflow is counting the RNA-seq reads for which we will use the [`featureCounts`](https://bio.tools/featurecounts) tool.
```

File: _episodes/04-reusing_tools.md [hunk 12]
```diff
@@ -56,4 +41,88 @@ To implement the last step the bio-cwl-tools library library will be used.
 {: .challenge}
 {% include links.md %}
 
+After finding the `featureCounts` tool, we need to download the tool to the directory in which our workflow is located.
+To download the tool, we go to the `featureCounts.cwl` page and open the raw file. 
+The screenshot below shows where you can find this button.
```
Comment (mr-c) [inline _episodes/04-reusing_tools.md]:
```suggestion
After finding the `featureCounts` tool, we need to download the tool to the directory in which our workflow is located.
To download the tool, we go to the `featureCounts.cwl` GitHub page and open the raw file. 
The screenshot below shows where you can find this button.
```

File: _episodes/04-reusing_tools.md [hunk 13]
```diff
@@ -56,4 +41,88 @@ To implement the last step the bio-cwl-tools library library will be used.
...
+
+!{}({{page.root}}/fig/screenshot_featureCounts_tool.png)
```
Comment (mr-c) [inline _episodes/04-reusing_tools.md]: This didn't work for me at https://deploy-preview-55--thirsty-hoover-1e0704.netlify.app/04-reusing_tools/index.html

File: _episodes/04-reusing_tools.md [hunk 14]
```diff
@@ -56,4 +41,88 @@ To implement the last step the bio-cwl-tools library library will be used.
...
+> >   quality_control:
+> >     run: bio-cwl-tools/fastqc/fastqc_2.cwl
+> > 	in:
+> > 	  reads_file: rna_reads_human
+> >     out: [html_file]
+> > 	
+> >   mapping_reads:
+> >     requirements:
+> >	      ResourceRequirement:
+> >	        ramMin: 9000
```
Comment (mr-c) [inline _episodes/04-reusing_tools.md]:
```suggestion
> >     requirements:
> >       ResourceRequirement:
> >         ramMin: 9000
```

removing more tabs

File: _episodes/04-reusing_tools.md [hunk 15]
```diff
@@ -11,9 +11,141 @@ keypoints:
...
+The last step of our workflow is counting the RNA-seq reads for which we will use the [`featureCounts`](https://bio.tools/featurecounts) tool.
+
+> ## Exercise
+>
+> Find the `featureCounts` tool in the [bio-cwl-tools library](https://github.com/common-workflow-library/bio-cwl-tools).
+> Have a look at the CWL document. Which inputs does this tool need? And what are the outputs of this tool?
+>
+> > ## Solution
+> > The `featureCounts` CWL document has 2 inputs: `annotations` and `mapped_reads`, both files. These inputs can be found on lines 6 and 9.
+> > The output of this tool is a file called `featurecounts` (line 21).
```
Comment (mr-c) [inline _episodes/04-reusing_tools.md]:
```suggestion
> > The `featureCounts` CWL document can be found at https://github.com/common-workflow-library/bio-cwl-tools/blob/release/subread/featureCounts.cwl ; it has 2 inputs: `annotations` (line 6) and `mapped_reads`, both files. These inputs can be found on lines 6 and 9.
> > The output of this tool is a file called `featurecounts` (line 21).
```

**Review by mr-c — CHANGES_REQUESTED** [review]
Thanks!

File: _episodes/04-reusing_tools.md (hunk 13 above)
Comment (ALuesink) [reply _episodes/04-reusing_tools.md]: Finally fixed in commit https://github.com/carpentries-incubator/cwl-novice-tutorial/pull/55/commits/028de017081dc5e0daa7adbc5c8e8ed11f305a20

**Review by mr-c — APPROVED** [review]

**Pull request merged**